File: src/main/java/edu/practice/datastructures/Stack.java
package edu.practice.datastructures;

import java.util.EmptyStackException;

public class Stack<E> {

    private static class Node<E> {
        E data;
        Node<E> next;

        Node(E data, Node<E> next) {
            this.data = data;
            this.next = next;
        }
    }

    private Node<E> top;
    private int size;

    public Stack() { // Initializes empty stack
        top = null;
        size = 0;
    }

    public int size() { return size; }

    public void push(E item) {
        top = new Node<>(item, top);
        size++;
    }

    public E peek() {
        if (isEmpty()) {
            throw new EmptyStackException();
        }
        return top.data;
    }

    public E pop() {
        E item;
        item = peek();

        top = top.next;
        size--;

        return item;
    }

    public boolean isEmpty() {
        return (size == 0);
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("TOP -> ");

        Node<E> current = top;
        while (current != null) {
            sb.append(current.data);

            if (current.next != null) {
                sb.append(" -> ");
            }

            current = current.next;
        }

        sb.append(" NULL");

        return sb.toString();
    }
}

File: src/main/kotlin/datastructures/Stack.kt
package edu.practice.datastructures

class Stack<E> {

    private data class Node<E>(val data: E, var next: Node<E>?)

    private var top: Node<E>? = null
    private var size: Int = 0

    fun size(): Int = size

    fun isEmpty(): Boolean = (size == 0)

    fun peek(): E {
        return top?.data ?: throw NoSuchElementException("Stack is Empty")
    }

    fun push(item: E) {
        top = Node(item, next = top)
        size++
    }

    fun pop(): E {
        val item = peek()
        top = top?.next
        size--
        return item
    }

    override fun toString(): String {
        val sb = StringBuilder("TOP")

        var current = top
        while (current != null) {
            sb.append(" -> ")
            sb.append(current.data)

            current = current.next
        }

        sb.append(" -> NULL")

        return sb.toString()
    }
}